pkg/client/request.go

```diff
@@ -619,8 +619,23 @@ func (r *Request) Upgrade(config *Config, newRoundTripperFunc func(*tls.Config)
 	return upgradeRoundTripper.NewConnection(resp)
 }
 
-// DoRaw executes a raw request which is not subject to interpretation as an API response.
-func (r *Request) DoRaw() ([]byte, error) {
+// request connects to the server and invokes the provided function when a server response is
+// received. It handles retry behavior and up front validation of requests. It wil invoke
+// fn at most once. It will return an error if a problem occured prior to connecting to the
+// server - the provided function is responsible for handling server errors.
+func (r *Request) request(fn func(*http.Request, *http.Response)) error {
+	if r.err != nil {
+		return r.err
+	}
+
+	// TODO: added to catch programmer errors (invoking operations with an object with an empty namespace)
+	if (r.verb == "GET" || r.verb == "PUT" || r.verb == "DELETE") && r.namespaceSet && len(r.resourceName) > 0 && len(r.namespace) == 0 {
+		return fmt.Errorf("an empty namespace may not be set when a resource name is provided")
```

Comment:
The client shouldn't set a namespace then (it shouldn't call client.Request().Namespace()) if the type doesn't have a namespace.  Basically we know whether namespace is set and empty vs unset and empty.

----- Original Message -----
> > @@ -619,8 +619,23 @@ func (r *Request) Upgrade(config *Config,
> > newRoundTripperFunc func(*tls.Config)
> >  	return upgradeRoundTripper.NewConnection(resp)
> >  }
> >  
> > -// DoRaw executes a raw request which is not subject to interpretation as
> > an API response.
> > -func (r *Request) DoRaw() ([]byte, error) {
> > +// request connects to the server and invokes the provided function when a
> > server response is
> > +// received. It handles retry behavior and up front validation of
> > requests. It wil invoke
> > +// fn at most once. It will return an error if a problem occured prior to
> > connecting to the
> > +// server - the provided function is responsible for handling server
> > errors.
> > +func (r *Request) request(fn func(*http.Request, *http.Response)) error {
> > +	if r.err != nil {
> > +		return r.err
> > +	}
> > +
> > +	// TODO: added to catch programmer errors (invoking operations with an
> > object with an empty namespace)
> > +	if (r.verb == "GET" || r.verb == "PUT" || r.verb == "DELETE") &&
> > r.namespaceSet && len(r.resourceName) > 0 && len(r.namespace) == 0 {
> > +		return fmt.Errorf("an empty namespace may not be set when a resource
> > name is provided")
> 
> I thought some objects (e.g. nodes, some events) didn't have a namespace?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/6671/files#r28164366